### ISSUE TITLE
Bug in NewFromFloat

### ISSUE THREAD

**Issue by wedancedalot** [shopspring/decimal#87]
Hi, after commit **78e9b82f68c54fcfaf34d71d751e594d20d2e242** NewFromFloat works not as it should

`    fmt.Println(decimal.NewFromFloat(0.1).String())
    fmt.Println(decimal.NewFromFloat(0.01).String())
    fmt.Println(decimal.NewFromFloat(0.02).String())
    fmt.Println(decimal.NewFromFloat(0.0023).String())
    fmt.Println(decimal.NewFromFloat(0.69).String())
    fmt.Println(decimal.NewFromFloat(0.98).String())`

return 

0.1000000000000000055511151231257827021181583404541015625
0.01000000000000000020816681711721685132943093776702880859375
0.0200000000000000004163336342344337026588618755340576171875
0.0022999999999999999611421941381195210851728916168212890625
0.689999999999999946709294817992486059665679931640625
0.979999999999999982236431605997495353221893310546875

Reverting back to commit **bf9a39e28bc9aea136acaf2c3628e68f65e3c3df** works as it should:
0.1
0.01
0.02
0.0023
0.69
0.98

Tested on go version go1.9.2 linux/amd64
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
Model name:            Intel(R) Core(TM) i3-4170 CPU @ 3.70GHz

**Comment by igrmk** [comment]
This is correct output.
Please use NewFromString or NewFromFloatWithExponent.

Here is pure Go output of your numbers.
https://play.golang.org/p/EwOamWFOBEA

**Comment by Nargott** [comment]
Are you kidding? Float is floating point type, but this library is "**Arbitrary-precision fixed-point** decimal numbers" which is described in your description.

**Comment by igrmk** [comment]
Nargott, I'm not kidding. If you write 0.1 it really means 0.1000000000000000055511151231257827021181583404541015625 because it is Float literal. Obviously if you initialize decimal from this number it should contain all of these digits. If you really need 0.1 you should use NewFromString or NewFromFloatWithExponent.

**Comment by wedancedalot** [comment]
Probably it'd be great to to let other developers know that library's behavior has changed this much.

**Comment by igrmk** [comment]
I agree. It doesn't look like minor change. I would issue new major version but it is up to mainteiners. I wrote pull request for this change. What do you think @victorquinn?

However expectation that decimal from 0.1 literal should be 0.1 is obviously a bug.

**Comment by igrmk** [comment]
BTW it may be worth to rename NewFromFloat to NewFromFloatExact. I guess it should help developers to think about it right way. There is an issue for such a change https://github.com/shopspring/decimal/issues/79.

**Comment by igrmk** [comment]
By renaming I mean deprecating and copying a method of course. But maybe it is better to do even a breaking change because previous version could lead to arbitrary loss of precision. It basically used FormatFloat with -1 precision that means shortest representation that is enough to construct a Float exactly. So for number 0.1000000000000000055511151231257827021181583404541015625 shortest representation is 0.1 effectively cutting the rest of the digits. So using NewFromFloat with previous version of this library is a bug by itself. And so forcing developers to review it may be a good thing.

**Comment by StanChe** [comment]
The biggest problem with such a change is that the previous code (supposedly bugged) got to the production. And with similar problems in other libraries (like storing/processing anything financial related in float64) it was the even bugs case negating each other. And the fix mentioned had broken the parity. So the renaming seems a much better way to avoid such problems.

**Comment by victorquinn** [comment]
Yowza, just seeing this now. 

The intended behavior was not for the library to change after 78e9b82f68c54fcfaf34d71d751e594d20d2e242 it was just to refactor and make it more DRY. And given the decent test coverage that seemed safe though it appears that it wasn't.

So if it did change behavior that's not good and I'd consider this a bug we should fix ASAP. Doing more investigation into the problem and solutions...

**Comment by victorquinn** [comment]
BTW until it's resolved you can fall back to the `1.0.0` tag which is before that change

**Comment by igrmk** [comment]
Hi @victorquinn. If you need some explanation about my change please feel free to ask. I will be available for a couple of hours from now today.

**Comment by victorquinn** [comment]
@igrmk maybe I misunderstood your change -- I thought 78e9b82f68c54fcfaf34d71d751e594d20d2e242 was to stop having 2 ways to create these numbers (`NewFromFloat` and `NewFromFloatWithExponent`) and leverage the same code for both. 

What I didn't expect was that it'd introduce breaking changes to the existing API and in the values it returned.

I am all for having better precision and so on, but changing the way a core function works in a way that breaks existing integrations is not acceptable.

Perhaps the best path forward here is to revert that change and add another method as you propose in https://github.com/shopspring/decimal/issues/79 which behaves as this code does. So to make it more clear/explicit:

1. Revert 78e9b82f68c54fcfaf34d71d751e594d20d2e242
1. Create a new method `NewFromFloatExact` (naming you suggest in #79 ) which behaves as `NewFromFloat` did after 78e9b82f68c54fcfaf34d71d751e594d20d2e242

Does that make sense @igrmk ? Basically I want the best way to both incorporate your change while not breaking backwards compatibility.

**Comment by igrmk** [comment]
I see. I had three depending pull requests that had a lot of bugfixes at first place.
1. They fixed NewFromFloatWithExponent and added several tests to check it.
  a. There was an issue with precision on rounding,
  b. It could not handle big numbers,
  c. Maybe some more but I cannot recall it from hand.
2.  They fixed NewFromFloat that I considered buggy because of precision loss and updated tests accordingly. From that moment it stored float exactly. And that's why people are complaining now. I guess I should explain this change better because I see misunderstanding now.
3. They added several additional tests for extreme values.

I think your suggestion makes a lot of sense.
Only I would mark NewFromFloat as deprecated and warn users about its usage.
Because relying on floating point value to have short decimal representation is really really bad.
Basically the people complaining about it now relied on a bug.
Obviously it could lead to much worse bug some day.

**Comment by igrmk** [comment]
One more thing. Reverting a change would reintroduce this one https://github.com/shopspring/decimal/issues/74.
It is opened but actually it is fixed with new algorithm.

**Comment by klauspost** [comment]
> that I considered buggy

@igrmk - just because you consider it buggy doesn't mean you can *just go breaking peoples production code* - especially with a package often used for things involving money.

While the original code definitely had its downsides, it worked in a specific way that you have now broken, making it impossible to upgrade. And what is making it worse is that nobody has done nothing to fix it for several months, so people now rely on *new behaviour*.

Wow. So glad this package isn't used in production code. (that was sarcasm in case you were wondering)

**Comment by klauspost** [comment]
Fixed by #100
